.circleci/config.yml

```diff
@@ -541,19 +541,27 @@ commands:
           name: Split tests
           working_directory: ~/project/e2e_tests
           command: |
-            # If a test mark is specified, preselect only files that contain tests with that mark to
-            # minimize how wrong CircleCI's splitting can get things.
-            if [ -n "<<parameters.mark>>" ]; then
-              mark="<<parameters.mark>>"
-              # Extract the first mark for complex conditions such as 'mark1 and not mark2'.
-              mark_prefix=${mark%% and*}
-              find tests -name 'test*.py' -print0 | xargs -0 grep -rl "pytest\.mark\.$mark_prefix"
-            else
-              circleci tests glob 'tests/**/test*.py'
-            fi | circleci tests split --split-by=timings > /tmp/tests-to-run
+            # Preselect the files that contain any matching tests to minimize how wrong CircleCI's
+            # splitting can get things. (The presence of files with no tests can cause the tests
+            # with files to be bunched up incorrectly, making the split suboptimal.)
+
+            # `--setup-plan` tells pytest to not run any tests but instead print out a description
+            # of the tests that would be run, along with some other stuff. Each file containing
+            # any selected tests is printed on a line by itself, except for (for some reason) a
+            # single trailing space.
+            pytest --setup-plan -m '<<parameters.mark>>' | grep '.py $' | sed 's/ $//' > /tmp/all-relevant-files
+
+            echo 'All files with tests matching mark "<<parameters.mark>>":'
+            sed 's/^/- /' </tmp/all-relevant-files
+
+            circleci tests split --split-by=timings < /tmp/all-relevant-files > /tmp/tests-to-run
             echo "Running tests from these files:"
             sed 's/^/- /' </tmp/tests-to-run
 
+            if ! [ -s /tmp/tests-to-run ]; then
```

Comment:
The negation outside of the test brackets is weird enough that now I have to ask "Why not `-z`?" :)

Reply:
lol seconded, but everything else here looks good

Reply:
I mean, I guess don't use `-z` if you actually want it to work. :stuck_out_tongue:  Doh.

I'm gonna argue that the `!` still belongs inside the brackets though. ;)

Reply:
Oh sure, I just didn't catch that you were referring to that/realize it could be done that way at all. Inside negation sounds good to me.